library/src/scala/internal/quoted/Matcher.scala

```diff
@@ -138,11 +139,28 @@ private[quoted] object Matcher {
             matched(scrutinee.seal)
 
           // Match a scala.internal.Quoted.patternHole and return the scrutinee tree
-          case (scrutinee: Term, TypeApply(patternHole, tpt :: Nil))
+          case (ClosedTerm(scrutinee), TypeApply(patternHole, tpt :: Nil))
               if patternHole.symbol == internal.Definitions_InternalQuoted_patternHole &&
                  scrutinee.tpe <:< tpt.tpe =>
             matched(scrutinee.seal)
 
+          // Matches an open term and wraps it into a lambda that provides the free variables
+          case (scrutinee, pattern @ Apply(Select(TypeApply(Ident("patternHole"), List(Inferred())), "apply"), args0 @ IdentArgs(args))) =>
+            def bodyFn(lambdaArgs: List[Tree]): Tree = {
+              val argsMap = args.map(_.symbol).zip(lambdaArgs.asInstanceOf[List[Term]]).toMap
+              new TreeMap {
+                override def transformTerm(tree: Term)(given ctx: Context): Term =
+                  tree match
+                    case tree: Ident => summon[Env].get(tree.symbol).flatMap(argsMap.get).getOrElse(tree)
+                    case tree => super.transformTerm(tree)
+              }.transformTree(scrutinee)
+            }
+            val names = args.map(_.name)
+            val argTypes = args0.map(x => x.tpe.widenTermRefExpr)
```

Comment:
Not sure `widenTermRefExpr` is safe, what if it's used as a singleton type inside?

Reply:
The singleton would not be phase consistent in the first place. It is always a reference to a symbol defined in the pattern.

Reply:
Agreed to keep it as it is, we see no way to break the code in reasonable usage.